### PR TITLE
New reflection primitive 'runSpeculative'

### PR DESCRIPTION
**Pull request by jespercockx**
Since the release is taking its time I would like to include the following new primitive which I have been using for a while for my WIP tactics library (https://github.com/jespercockx/ataca). It doesn't touch any existing code so it's unlikely to break anything, and it would allow other people to more easily experiment with tactics in Agda too.

### CODE REVIEW

File: src/full/Agda/TypeChecking/Unquote.hs [hunk 1]
```diff
@@ -734,3 +735,19 @@ evalTCM v = do
...
       primUnitUnit
+
+    tcRunSpeculative :: Term -> UnquoteM Term
+    tcRunSpeculative mu = do
+      oldState <- getTC
+      u <- reduce =<< evalTCM mu
+      true  <- liftTCM primTrue
+      false <- liftTCM primFalse
+      case u of
+        Con _ _ [Apply (Arg { unArg = x }), Apply (Arg { unArg = b0 })] -> do
```
Comment (UlfNorell) [inline src/full/Agda/TypeChecking/Unquote.hs]: I think you could use `unquote u :: UnquoteM Bool` here.

**Review by UlfNorell — APPROVED** [review]

**Comment by guillaumebrunerie** [comment]
The new primitive should also be added to the documentation. There is a very helpful list of all primitives (https://agda.readthedocs.io/en/latest/language/reflection.html#type-checking-computations), it would be a shame to forget this one.

File: src/full/Agda/TypeChecking/Unquote.hs [hunk 2]
```diff
@@ -734,3 +735,14 @@ evalTCM v = do
...
+    tcRunSpeculative :: Term -> UnquoteM Term
+    tcRunSpeculative mu = do
+      oldState <- getTC
+      u <- reduce =<< evalTCM mu
+      case u of
+        Con _ _ [Apply (Arg { unArg = x }), Apply (Arg { unArg = b })] -> do
+          unlessM (unquote b) $ putTC oldState
+          return x
+        _ -> liftTCM $ typeError . GenericDocError =<<
+          "Should be a pair: " <+> prettyTCM u
```
Comment (gallais) [inline src/full/Agda/TypeChecking/Unquote.hs:748]: Given the type of the `BUILTIN`, shouldn't this be an `__IMPOSSIBLE__`?

Comment (jespercockx) [reply src/full/Agda/TypeChecking/Unquote.hs:748]: If the user is silly, it could be a postulate or something..

**Pull request merged**